crates/bevy_state/src/state/state_set.rs

```diff
@@ -140,45 +141,70 @@ impl<S: InnerStateSet> StateSet for S {
+        // | false          | false      | true           | -            | -                                |
+        // | false          | true       | false          | false        | -                                |
+        // | true           | false      | false          | false        | -                                |
+        // | true           | true       | false          | false        | -                                |
+        // | true           | false      | true           | false        | Some(current) -> None            |
+        // | true           | true       | true           | false        | Some(current) -> None            |
+        // | true           | false      | false          | true         | None -> Some(default)            |
+        // | true           | true       | false          | true         | None -> Some(next)               |
+        // | true           | true       | true           | true         | Some(current) -> Some(next)      |
+        // | false          | true       | true           | true         | Some(current) -> Some(next)      |
+        // | true           | false      | true           | true         | Some(current) -> Some(current)   |
+
+        let apply_state_transition =
+            |mut parent_changed: EventReader<StateTransitionEvent<S::RawState>>,
+             event: EventWriter<StateTransitionEvent<T>>,
+             commands: Commands,
+             current_state_res: Option<ResMut<State<T>>>,
+             next_state_res: Option<ResMut<NextState<T>>>,
+             state_set: Option<Res<State<S::RawState>>>| {
+                let parent_changed = parent_changed.read().last().is_some();
+                let next_state = take_next_state(next_state_res);
+
+                if !parent_changed && next_state.is_none() {
+                    return;
+                }
+
+                let current_state = current_state_res.as_ref().map(|s| s.get()).cloned();
 
-            let new_state =
-                if let Some(state_set) = S::convert_to_usable_state(state_set.as_deref()) {
-                    T::should_exist(state_set)
-                } else {
-                    None
+                let should_exist = match parent_changed {
+                    true => {
+                        if let Some(state_set) = S::convert_to_usable_state(state_set.as_deref()) {
+                            T::should_exist(state_set)
+                        } else {
+                            None
+                        }
+                    }
+                    false => current_state.clone(),
                 };
 
-            match new_state {
-                Some(value) => {
-                    if current_state.is_none() {
+                match should_exist {
+                    Some(initial_state) => {
+                        let new_state = match (current_state, next_state) {
```

Comment:
This could also be written as `let new_state = next_state.or(current_state).unwrap_or(initial_state);` :)